File: benchmark/src/perl/linked_list.pl
use Time::HiRes qw(time);
my $N=100000;my $seed=12345;my $acc=0;my $start=time();
my $head=undef;for my $i (1..$N){$seed=($seed*16807)%2147483647;$head={value=>$seed%10000,next=>$head}}
my($prev,$cur)=(undef,$head);for my $i (1..$N){my $nxt=$cur->{next};$cur->{next}=$prev;$prev=$cur;$cur=$nxt}
my($pos,$walk)=(0,$prev);for my $i (1..$N){$acc=($acc+$walk->{value}*$pos)%1000003;$pos++;$walk=$walk->{next}}
my $end=time();my $ms=($end-$start)*1000;print "Result: $acc\n";printf "Time: %.0fms\n",$ms;

File: benchmark/src/ruby/linked_list.rb
N=100000;seed=12345;acc=0;start=Time.now.to_f*1000
head=nil;N.times{seed=(seed*16807)%2147483647;head={value:seed%10000,next:head}}
prev=nil;cur=head;N.times{nxt=cur[:next];cur[:next]=prev;prev=cur;cur=nxt}
pos=0;walk=prev;N.times{acc=(acc+walk[:value]*pos)%1000003;pos+=1;walk=walk[:next]}
finish=Time.now.to_f*1000;puts "Result: #{acc}";puts "Time: #{(finish-start).round}ms"
